src/core/utils/ledger.ts

```diff
@@ -0,0 +1,240 @@
+        }
+
+        let response = Buffer.alloc(0);
+        for (let apdu of apdus) {
+            response = await this.transport.send(apdu.cla, apdu.ins, apdu.p1, apdu.p2, apdu.data);
+        }
+        //Response from Ledger
+        let h = response.toString("hex");
+
+        let signature = h.slice(0, 128);
+
+        let payload = rawPayload.slice(0, 16) + 
+            signature + 
+            signer + 
+            rawPayload.slice(16 + 128 + 64, rawPayload.length);
+
+        let generationHashBytes = Array.from(Convert.hexToUint8(networkGenerationHash));
+
+        let transactionHash = Transaction.createTransactionHash(
+            payload,
+            generationHashBytes,
+            transferTransaction.networkType
+        );
+
+        let signedTransaction = new SignedTransaction(
+            payload,
+            transactionHash,
+            signer,
+            transferTransaction.type,
+            transferTransaction.networkType
+        );
+
+            return signedTransaction
+    }
+    async signCosignatureTransaction(path: string, transferTransaction:AggregateTransaction, networkGenerationHash: string, signer: string) {
+        const bipPath = BIPPath.fromString(path).toPathArray();
+        const rawPayload = transferTransaction.serialize();
+        const signingBytes =  transferTransaction.transactionInfo.hash + rawPayload.slice(216);
+        const rawTx = Buffer.from(signingBytes, "hex");
+        const curveMask = 0x80;
+        //symbol-sdk 0.17.3
+
+        const apdus = [];
+        let offset = 0;
+        let twiceTransfer;
+        //The length of the APDU buffer is 255Bytes
+        if (rawTx.length > 446) {
+            throw new Error("The transaction is too long.");
+        } else {
+            twiceTransfer = rawTx.length > 234 ? true : false;
+        }
+
+        while (offset !== rawTx.length) {
+            const maxChunkSize = offset === 0 ? MAX_CHUNK_SIZE - 1 - bipPath.length * 4 : MAX_CHUNK_SIZE;
+            const chunkSize = offset + maxChunkSize > rawTx.length ? rawTx.length - offset : maxChunkSize;
+            const apdu = {
+                cla: 0xe0,
+                ins: 0x04,
+                p1: offset === 0 ? 0x00 : 0x80,
+                p2: curveMask,
+                data: offset === 0 ? Buffer.alloc(1 + bipPath.length * 4 + chunkSize) : Buffer.alloc(chunkSize)
+            };
+
+            if (offset === 0) {
+                apdu.data.writeInt8(bipPath.length, 0);
+                bipPath.forEach((segment, index) => {
+                    apdu.data.writeUInt32BE(segment, 1 + index * 4);
+                });
+                rawTx.copy(apdu.data, 1 + bipPath.length * 4, offset, offset + chunkSize);
+                if (!twiceTransfer) {
+                    apdu.p1 = 0x90;
+                }
+            } else {
+                rawTx.copy(apdu.data, 0, offset, offset + chunkSize);
+            }
+
+            apdus.push(apdu);
+            offset += chunkSize;
```

Comment:
this can probably be refactored to `generateDataUnit` or `getDataUnitBytes`

`APDU` wording is complex and structure of data is not clear enough to keep this here.